cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,260 @@ where
+    // | st2g(tagged_ptr, iter_ptr + offset + 32)          | repeat `n` times, i.e. while (remaining_bits >= 32) holds
+    // | ...                                               |
+    // | st2g(tagged_ptr, iter_ptr + offset + (n-1) * 32) /
+    // |
+    // | stg(tagged_ptr, iter_ptr + offset + n * 32)  | generate only if required
+    // +---
+    //
+    // +---next_block
+    // +---
+
+    // TODO: only use one of i64 or u64
+
+    // If the number of bytes to be tagged is larger than this threshhold,
+    // insert a loop that tags using `st2g` chunks.
+    // Threshhold should be a multiple of 32, so each `st2g` chunk can be
+    // exactly divided into only 32-byte tagging `st2g` instructions
+    // This constant was adapted from clang.
+    const LOOP_UNROLL_BYTES_THRESHHOLD_I64: i64 = 160;
+    const LOOP_UNROLL_BYTES_THRESHHOLD_U64: u64 = 160;
+
     // Assert that `size` is 16 byte aligned, and trap if not
     if size % 16 != 0 {
         return Err(WasmError::User(format!(
             "size must be 16 byte aligned, was {size}"
         )));
     }
 
-    // First insert an overflow check. This allows us to insert non-trapping instructions later.
-    let c = builder.ins().iconst(I64, size);
+    // TODO: maybe optimize by adding `size` statically as immediate
+    // Trap if (iter_ptr + size) overflows, so we can use non-trapping instructions later
     builder
-    .ins()
-    .uadd_overflow_trap(iter_ptr, c, ir::TrapCode::HeapOutOfBounds);
+        .ins()
+        .uadd_overflow_trap(iter_ptr, size_dynamic, ir::TrapCode::HeapOutOfBounds);
 
-    let mut i = 0;
-    while size - i >= 32 {
-        let iter_ptr = builder.ins().iadd_imm(iter_ptr, i);
-        builder.ins().arm64_st2g(tagged_ptr, iter_ptr);
-        i += 32;
-    }
-    if i < size {
-        let iter_ptr = builder.ins().iadd_imm(iter_ptr, i);
-        builder.ins().arm64_stg(tagged_ptr, iter_ptr);
+    let size_u64: u64 = size as u64;
+    // In Rust, for u64: (a / b) == floor(a / b), which we make use of here
+    let num_required_st2g_loops: u64 = size_u64 / LOOP_UNROLL_BYTES_THRESHHOLD_U64;
+    let num_required_last_stgs: u64 = (size_u64 % LOOP_UNROLL_BYTES_THRESHHOLD_U64) / 16;
+
+    // TODO: cast may be wrong, but i64 is required for use as a Imm64
+    let num_required_st2g_loops_i64: i64 = num_required_st2g_loops as i64;
+
+    if num_required_st2g_loops <= 1 {
+        // WITHOUT_LOOP_UNROLLING_THRESHHOLD
+
+        tag_memory_region_static_without_loop_unrolling_threshhold(
+            iter_ptr, 0, tagged_ptr, size, builder,
+        );
+    } else {
+        // WITH_LOOP_UNROLLING_THRESHHOLD
+
+        // === Block definitions
+        // st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+        let st2g_loop_condition_block =
+            block_with_params(builder, [ValType::I64, ValType::I64], environ)?;
+
+        // TODO: does this really require iter_ptr and counter as parameters, since they are never modified in the condition
```

Comment:
I think these parameters are required, because the values will have been changed by another block when we use them here, hence we need to read them.

Reply:
Yes. I like the approach that you described in the comment above, where you would just increment `iter_ptr` and use `iter_ptr` in your condition. In that case you can remove one of the parameters.